infer/inference.cc

```diff
@@ -269,6 +269,14 @@ unique_ptr<cfg::CFG> Inference::run(core::Context ctx, unique_ptr<cfg::CFG> cfg)
         for (auto &pair : current.vars()) {
             pair.second.knowledge.emitKnowledgeSizeMetric();
         }
+
+        core::TypePtr conditionalType = current.getTypeAndOrigin(ctx, bb->bexit.cond.variable).type;
+
+        if (!conditionalType.isUntyped() && core::Types::isSubType(ctx, core::Types::void_(), conditionalType)) {
+            if (auto e = ctx.beginError(bb->bexit.loc, core::errors::Infer::InvalidVoidUsage)) {
+                e.setHeader("Can't use void types in conditional");
```

Comment:
Can we make this error message a little better by using:

```cpp
e.addErrorSection(conditionalType.explainGot(gs, originForUninitialized));
```

?

Reply:
That looks better, I added the section.

Some tests were giving false positives when the type of an argument was `Object`, which seemed to match the condition of having `void` as a subtype. I added a check to prevent that if the type is object.

Here's what the output looks like now.

<img width="751" alt="Screen Shot 2021-06-22 at 9 14 57 AM" src="https://user-images.githubusercontent.com/18742907/122931279-bac57a00-d33a-11eb-9452-8f9a0e1736d4.png">

Reply:
> Some tests were giving false positives when the type of an argument was `Object`, which seemed to match the condition of having `void` as a subtype. I added a check to prevent that if the type is object.

Interesting. I wonder if this means that we should make void be a subtype of `<top>` rather than `Object`.

> Here's what the output looks like now.

🤔 i'm surprised to see it say `NilClass` instead of `Void`.

Reply:
I was surprised by that too. Could it be because the method body is empty (thus returning `nil`)?

Reply:
Actually, I just tested this and it also gives us false positives for `BasicObject`. I can add another check for that for the time being.

Reply:
Added a new check for `BasicObject`, which now passes in our code base.